[mono] Re-enable tests that call MakeGenericType with non-RuntimeType arguments

lambdageek opened the pull request:
Partially fixes https://github.com/dotnet/runtime/issues/32743

The actual fix was in https://github.com/dotnet/runtime/pull/58014

Some of the re-enabled tests failed on wasm due to https://github.com/dotnet/runtime/issues/52062, so those have been disabled here.

lambdageek committed:
[mono] Re-enable tests that call MakeGenericType with non-RuntimeType arguments

Fixes https://github.com/dotnet/runtime/issues/32743

The actual fix was in https://github.com/dotnet/runtime/pull/58014

lambdageek commented:
Replaces #64195

radical commented:
/azp run runtime-extra-platforms

lambdageek commented:
Some failures on wasm and tvOS.  In both cases it looks like the tests probably need to be disabled.

Wasm example. Looks like it's trying to find an assembly in the file system, which won't work:
```
[18:08:27] fail: [FAIL] Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.WithNullLevel_GeneratorWontFail(argumentList: "EventId = 0, Level = null, Message = \"This is a m"...)
[18:08:27] info: System.ArgumentException : Empty path name is not legal. (Parameter 'path')
[18:08:27] info:    at System.IO.Strategies.FileStreamHelpers.ValidateArguments(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, Int64 preallocationSize)
[18:08:27] info:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, Int64 preallocationSize)
[18:08:27] info:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
[18:08:27] info:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, Boolean useAsync)
[18:08:27] info:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
[18:08:27] info:    at Roslyn.Utilities.StandardFileSystem.OpenFile(String filePath, FileMode mode, FileAccess access, FileShare share)
[18:08:27] info:    at Roslyn.Utilities.CommonCompilerFileSystemExtensions.OpenFileWithNormalizedException(ICommonCompilerFileSystem fileSystem, String filePath, FileMode fileMode, FileAccess fileAccess, FileShare fileShare)
[18:08:27] info:    at Microsoft.CodeAnalysis.MetadataReference.CreateFromFile(String path, MetadataReferenceProperties properties, DocumentationProvider documentation)
[18:08:27] info:    at SourceGenerators.Tests.RoslynTestUtils.CreateTestProject(IEnumerable`1 references, Boolean includeBaseReferences)
[18:08:27] info:    at SourceGenerators.Tests.RoslynTestUtils.RunGenerator(ISourceGenerator generator, IEnumerable`1 references, IEnumerable`1 sources, Boolean includeBaseReferences, CancellationToken cancellationToken)
[18:08:27] info:    at Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.RunGenerator(String code, Boolean wrap, Boolean inNamespace, Boolean includeBaseReferences, Boolean includeLoggingReferences, CancellationToken cancellationToken)
[18:08:27] info:    at Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.WithNullLevel_GeneratorWontFail(String argumentList)
[18:08:27] info: --- End of stack trace from previous location ---
```

tvOS example. `MakeGenericType` with non-`RuntimeType` arguments isn't supported in FullAOT:

```
System.PlatformNotSupportedException : Operation is not supported on this platform.

Stack trace
   at System.RuntimeType.MakeGenericType(Type[] instantiation)
   at System.ComponentModel.Composition.Hosting.CompositionServices.AdjustSpecifiedTypeIdentityType(Type specifiedContractType, Type memberType)
   at System.ComponentModel.Composition.Hosting.CompositionServices.AdjustSpecifiedTypeIdentityType(Type specifiedContractType, MemberInfo member)
   at System.ComponentModel.Composition.Hosting.CompositionServices.GetTypeIdentityTypeFromExport(MemberInfo member, ExportAttribute export)
   at System.ComponentModel.Composition.Hosting.CompositionServices.GetContractInfoFromExport(MemberInfo member, ExportAttribute export, Type& typeIdentityType, String& contractName)
   at System.ComponentModel.Composition.AttributedModel.AttributedPartCreationInfo.CreateExportDefinition(MemberInfo member, ExportAttribute exportAttribute)
   at System.ComponentModel.Composition.AttributedModel.AttributedPartCreationInfo.GetExportDefinitions()
   at System.ComponentModel.Composition.AttributedModel.AttributedPartCreationInfo.DiscoverExportsAndImports()
   at System.ComponentModel.Composition.AttributedModel.AttributedPartCreationInfo.GetExports()
   at System.ComponentModel.Composition.ReflectionModel.ReflectionComposablePartDefinition.get_ExportDefinitionsInternal()
   at System.ComponentModel.Composition.ReflectionModel.ReflectionComposablePartDefinition.get_ExportDefinitions()
   at System.ComponentModel.Composition.Hosting.TypeCatalog.CreateIndex()
   at System.Lazy`1[[System.Collections.Generic.IDictionary`2[[System.String, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Collections.Generic.List`1[[System.ComponentModel.Composition.Primitives.ComposablePartDefinition, System.ComponentModel.Composition, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1[[System.Collections.Generic.IDictionary`2[[System.String, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Collections.Generic.List`1[[System.ComponentModel.Composition.Primitives.ComposablePartDefinition, System.ComponentModel.Composition, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1[[System.Collections.Generic.IDictionary`2[[System.String, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Collections.Generic.List`1[[System.ComponentModel.Composition.Primitives.ComposablePartDefinition, System.ComponentModel.Composition, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].CreateValue()
   at System.Lazy`1[[System.Collections.Generic.IDictionary`2[[System.String, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Collections.Generic.List`1[[System.ComponentModel.Composition.Primitives.ComposablePartDefinition, System.ComponentModel.Composition, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].get_Value()
   at System.ComponentModel.Composition.Hosting.TypeCatalog.GetCandidateParts(String contractName)
   at System.ComponentMo
```

radical commented:
The wasm ones are failing to get the assembly location with `string corelib = Assembly.GetAssembly(typeof(object))!.Location;`.
Issue: https://github.com/dotnet/runtime/issues/52062

radical committed:
[wasm] Disable some logging generator tests

`Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorEmitterTests.*`
`Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.*`

.. failing like:

```
[18:09:38] fail: [FAIL] Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial
[18:09:38] info: System.ArgumentException : Empty path name is not legal. (Parameter 'path')
[18:09:38] info:    at System.IO.Strategies.FileStreamHelpers.ValidateArguments(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, Int64 preallocationSize)
[18:09:38] info:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, Int64 preallocationSize)
[18:09:38] info:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
[18:09:38] info:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, Boolean useAsync)
[18:09:38] info:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
[18:09:38] info:    at Roslyn.Utilities.StandardFileSystem.OpenFile(String filePath, FileMode mode, FileAccess access, FileShare share)
[18:09:38] info:    at Roslyn.Utilities.CommonCompilerFileSystemExtensions.OpenFileWithNormalizedException(ICommonCompilerFileSystem fileSystem, String filePath, FileMode fileMode, FileAccess fileAccess, FileShare fileShare)
[18:09:38] info:    at Microsoft.CodeAnalysis.MetadataReference.CreateFromFile(String path, MetadataReferenceProperties properties, DocumentationProvider documentation)
[18:09:38] info:    at SourceGenerators.Tests.RoslynTestUtils.CreateTestProject(IEnumerable`1 references, Boolean includeBaseReferences)
[18:09:38] info:    at SourceGenerators.Tests.RoslynTestUtils.RunGenerator(ISourceGenerator generator, IEnumerable`1 references, IEnumerable`1 sources, Boolean includeBaseReferences, CancellationToken cancellationToken)
[18:09:38] info:    at Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.RunGenerator(String code, Boolean wrap, Boolean inNamespace, Boolean includeBaseReferences, Boolean includeLoggingReferences, CancellationToken cancellationToken)
[18:09:38] info:    at Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial()
[18:09:38] info: --- End of stack trace from previous location ---
```

.. due to https://github.com/dotnet/runtime/issues/52062 .

radical committed:
[tvos] Disable some failing tests

`System.ComponentModel.Composition.Registration.Tests.RegistrationBuilderTests.ExplicitGenericExportInRegistrationBuilder`
`System.ComponentModel.Composition.Registration.Tests.RegistrationBuilderTests.ExplicitGenericArity2ExportInRegistrationBuilder`
`System.ComponentModel.Composition.Registration.Tests.RegistrationBuilderTests.GenericBaseClassExportInRegistrationBuilder`
`System.ComponentModel.Composition.Registration.Tests.RegistrationBuilderTests.GenericInterfaceExportInRegistrationBuilder`
`System.ComponentModel.Composition.Registration.Tests.RegistrationBuilderTests.ShouldSucceed`

.. failing due to https://github.com/dotnet/runtime/issues/32743 .

radical commented:
/azp run runtime-extra-platforms

lambdageek commented on src/libraries/System.ComponentModel.Composition.Registration/tests/System/ComponentModel/Composition/Registration/RegistrationBuilderTests.cs:
```diff
@@ -45,7 +45,7 @@ public DiscoveredCatalog()
 
         [Fact]
-        [ActiveIssue("https://github.com/dotnet/runtime/issues/32743", TestRuntimes.Mono)]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/32743", TestPlatforms.tvOS)]
```
@radical I don't think this is an ActiveIssue.  (or at least 32743 is misleading).  I don't think we can support MakeGenericType for non-RuntimeType types in FullAOT (without interpreter fallback).

I think this should be `[ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotMonoAot))]`

marek-safar replied:
Yep, this is not tvOS specific. Using `RuntimeFeature.IsDynamicCode` or whatever flag NativeAOT uses would be better

/cc @jkotas

lambdageek replied:
I guess that's

https://github.com/dotnet/runtime/blob/3e4c4a4c2c3343ac545bf48b9c3189556bfecf73/src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs#L109

```
[ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsReflectionEmitSupported)]
```

Which brings up the question: do we run these tests with NativeAOT? I would expect them to fail there, too

lambdageek committed:
Disable RegistrationBuilderTests that require ref emit on platforms that don't support it

radical commented:
/azp run runtime-extra-platforms

lambdageek commented:
Failures are https://github.com/dotnet/runtime/issues/64389

MaximLipnin commented:
The re-enabled `Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorEmitterTests` tests started failing with PNSE on Mono+Windows legs.
